verification/io/router.gobra

```diff
@@ -22,34 +22,89 @@ package io
 
 // This function corresponds to function 'core' in the IO spec
 ghost
+requires dp.Valid()
 decreases
-pure func (dp DataPlaneSpec) core() set[IO_as]
+pure func (dp DataPlaneSpec) core() set[IO_as] {
+	return dp.GetCoreAS()
+}
+
+ghost 
+decreases
+pure func if2term(ifs option[IO_ifs]) IO_msgterm {
+	return match ifs {
+		case none[IO_ifs]: 
+			MsgTerm_Empty{}
+		default: 
+			IO_msgterm(MsgTerm_AS{IO_as(get(ifs))})
+	}
+}
+
+ghost
+decreases
+pure func (dp DataPlaneSpec) hf_valid(d bool, ts uint, uinfo set[IO_msgterm], hf IO_HF) bool {
+	return let inif := hf.InIF2 in
+		let egif := hf.EgIF2 in
+		let x := hf.HVF in
+		let l := IO_msgterm(MsgTerm_L{
+			seq[IO_msgterm]{
+				IO_msgterm(MsgTerm_Num{ts}), 
+				if2term(inif), 
+				if2term(egif), 
+				IO_msgterm(MsgTerm_FS{uinfo})}}) in 
+		x == mac(macKey(asidToKey(dp.asid())), l)
+}
 
 ghost
 decreases
-pure func (dp DataPlaneSpec) hf_valid(p1 bool, p2 uint, p3 set[IO_msgterm], p4 IO_HF) bool
+pure func macKey(key IO_key) IO_msgterm {
+	return IO_msgterm(MsgTerm_Key{key})
+}
+
+ghost 
+decreases
+pure func mac(fst IO_msgterm, snd IO_msgterm) IO_msgterm {
+	return IO_msgterm( MsgTerm_Hash { 
+		MsgTerm_Hash_ : IO_msgterm( MsgTerm_MPair{
+			MsgTerm_MPair_1 : fst,
+			MsgTerm_MPair_2 : snd, 
+		}),
+	})
+}
 
+// helper func
 ghost
 decreases
-pure func (dp DataPlaneSpec) upd_uinfo(p1 set[IO_msgterm], p2 IO_HF) set[IO_msgterm]
+pure func asidToKey(asid IO_as) IO_key{
+	return IO_key(Key_macK{asid})
+}
 
 ghost
 decreases
-pure func (dp DataPlaneSpec) rid() IO_rid
+pure func upd_uinfo(segid set[IO_msgterm], hf IO_HF) set[IO_msgterm]{
+	return let setHVF := set[IO_msgterm]{hf.HVF} in
+		(segid union setHVF) setminus (segid intersection setHVF)
+}
 
 ghost
 decreases
 pure func (dp DataPlaneSpec) asid() IO_as {
-	return dp.rid_as(dp.rid())
+	return dp.GetLocalIA()
 }
```

Comment:
This function is also redundant. We could rename `GetLocalIA` to `ASID` and use that instead everywhere in the IO spec

Reply:
I agree. However, `GetLocalIA()` is a function we defined for our DataplaneSpec while `asid()` is the original function from Isabelle. I thought it would be cleaner this way. Should I remove it?

Reply:
same as before